servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/HttpRequestAutoDrainingServiceFilter.java

```diff
@@ -0,0 +1,161 @@
+import io.servicetalk.http.api.HttpRequest;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+/**
+ * Configure automatic consumption of request {@link StreamingHttpRequest#payloadBody() payload body} when it is not
+ * consumed by the service.
+ * <p>
+ * For <a href="https://tools.ietf.org/html/rfc7230#section-6.3">persistent HTTP connections</a> it is required to
+ * eventually consume the entire request payload to enable reading of the next request. This is required because
+ * requests are pipelined for HTTP/1.1, so if the previous request is not completely read, next request can not be
+ * read from the socket. For cases when there is a possibility that user may forget to consume request payload,
+ * ServiceTalk automatically consumes request payload body.
+ * An example of guaranteed consumption are {@link HttpRequest non-streaming APIs}.
+ *
+ * @see io.servicetalk.http.api.HttpServerBuilder#drainRequestPayloadBody(boolean)
+ */
+public final class HttpRequestAutoDrainingServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    /**
+     * Singleton instance of the draining service filter.
+     */
+    public static final StreamingHttpServiceFilterFactory INSTANCE = new HttpRequestAutoDrainingServiceFilter();
+
+    private HttpRequestAutoDrainingServiceFilter() {
+        // singleton
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            final DrainTerminalSignalConsumer terminalSignalConsumer = new DrainTerminalSignalConsumer(request);
+            request.transformMessageBody(body ->
+                    body.beforeSubscriber(() -> {
+                        terminalSignalConsumer.requestSubscribed();
```

Comment:
(optional) we can avoid allocation of a captured lambda if `DrainTerminalSignalConsumer` also implements `Supplier`